pupil_src/shared_modules/marker_tracker_3d/controller/controller.py

```diff
@@ -0,0 +1,93 @@
+import logging
+
+from marker_tracker_3d import worker, controller
+from observable import Observable
+
+logger = logging.getLogger(__name__)
+
+
+class Controller(Observable):
+    def __init__(
+        self, controller_storage, model_storage, camera_intrinsics, task_manager, plugin
+    ):
+        self._controller_storage = controller_storage
+        self._model_storage = model_storage
+        self._camera_intrinsics = camera_intrinsics
+        self._plugin = plugin
+
+        self._visibility_graphs = worker.VisibilityGraphs(model_storage)
+        self._prepare_for_model_update = worker.PrepareForModelUpdate(
+            model_storage, predetermined_origin_marker_id=None
+        )
+        self._model_initialization_controller = controller.ModelInitializationController(
+            camera_intrinsics, task_manager
+        )
+        self._model_optimization_controller = controller.ModelOptimizationController(
+            model_storage, camera_intrinsics, task_manager
+        )
+        self._update_model_storage = worker.UpdateModelStorage(model_storage)
+
+        self._plugin.add_observer("recent_events", self._on_recent_events)
+        self._setup_model_update_pipeline()
+
+    def _setup_model_update_pipeline(self):
+        self._controller_storage.add_observer(
+            "update", self._visibility_graphs.add_observations
+        )
+        self._visibility_graphs.add_observer(
+            "on_novel_markers_added", self._prepare_for_model_update.run
+        )
+        self._prepare_for_model_update.add_observer(
+            "on_prepare_for_model_init_done", self._on_update_model_started
+        )
+        self.add_observer(
+            "_on_update_model_started", self._model_initialization_controller.run
+        )
+        self._model_initialization_controller.add_observer(
+            "on_model_init_done", self._model_optimization_controller.run
+        )
+        self._model_optimization_controller.add_observer(
+            "on_model_opt_done", self._update_model_storage.run
+        )
+        self._update_model_storage.add_observer(
+            "on_update_model_storage_done", self._on_update_model_finished
+        )
+        self.add_observer(
+            "_on_update_model_finished", self._prepare_for_model_update.run
+        )
+
+    def _update(self, frame):
+        marker_id_to_detections = worker.detect_markers.detect(
+            frame, self._controller_storage.min_marker_perimeter
+        )
+        camera_extrinsics = worker.localize_camera.localize(
+            self._camera_intrinsics,
+            marker_id_to_detections,
+            self._model_storage.marker_id_to_extrinsics_opt,
+            camera_extrinsics_prv=self._controller_storage.camera_extrinsics,
+        )
+        self._controller_storage.update(marker_id_to_detections, camera_extrinsics)
+
+    def _on_update_model_started(self, args=None):
```

Comment:
`args` is usually only used when you want to take any number of arguments, i.e. as `*args` (please ask if you don't know what I mean). In your case, you could just go with `_`, which would clearer (you expect one argument, but do not care about it). The `=None` is not necessary for you (as far as I can see) as this is never called without the single argument `data_for_model_init`